[C,lib/utils/string.js,107]         }
[C,lib/utils/string.js,108]       }
[C,lib/utils/string.js,109] 
[C,lib/utils/string.js,110]       return '{' + entries.join(', ') + '}';
[C,lib/utils/string.js,111]     }
[C,lib/utils/string.js,112]   }
[C,lib/utils/string.js,113] 
[C,lib/utils/string.js,114]   return String(value);
[C,lib/utils/string.js,115] };
[C,lib/utils/string.js,116] 
[A,lib/utils/string.js,117] /**
[A,lib/utils/string.js,118]  * Stringify a value into a string enclosed in double quotes.
[A,lib/utils/string.js,119]  * Double quotes inside the value are escaped.
[A,lib/utils/string.js,120]  * @param {*} value
[A,lib/utils/string.js,121]  * @return {string}
[A,lib/utils/string.js,122]  */
[A,lib/utils/string.js,123] exports.stringify = function (value) {
[A,lib/utils/string.js,124]   var str = String(value).replace(/([^\\]|^)"/g, function (full, before) {
[A,lib/utils/string.js,125]     return before + '\\"';
[A,lib/utils/string.js,126]   });
[A,lib/utils/string.js,127] 
[A,lib/utils/string.js,128]   return '"' + str + '"';
[A,lib/utils/string.js,129] }
[A,lib/utils/string.js,130] 
[C,lib/utils/string.js,131] /**
[C,lib/utils/string.js,132]  * Recursively format an n-dimensional matrix
[C,lib/utils/string.js,133]  * Example output: "[[1, 2], [3, 4]]"
[C,lib/utils/string.js,134]  * @param {Array} array
[C,lib/utils/string.js,135]  * @param {Object | number | Function} [options]  Formatting options. See
[C,lib/utils/string.js,136]  *                                                lib/utils/number:format for a
[C,lib/utils/string.js,137]  *                                                description of the available
[C,lib/utils/string.js,138]  *                                                options.
[C,lib/utils/string.js,139]  * @returns {string} str
[C,lib/utils/string.js,140]  */
[C,lib/utils/string.js,141] function formatArray (array, options) {
[C,lib/utils/string.js,142]   if (Array.isArray(array)) {
[C,lib/utils/string.js,143]     var str = '[';
[C,lib/utils/string.js,144]     var len = array.length;
[C,lib/utils/string.js,145]     for (var i = 0; i < len; i++) {
@@ -68,30 +68,54 @@ describe('ConstantNode', function() {
[C,test/expression/node/ConstantNode.test.js,68]     assert.strictEqual(expr.eval(), null);
[C,test/expression/node/ConstantNode.test.js,69] 
[C,test/expression/node/ConstantNode.test.js,70]   });
[C,test/expression/node/ConstantNode.test.js,71] 
[C,test/expression/node/ConstantNode.test.js,72]   it ('should compile a ConstantNode with bigmath', function () {
[C,test/expression/node/ConstantNode.test.js,73]     var expr = new bigmath.expression.node.ConstantNode('2.3', 'number').compile();
[C,test/expression/node/ConstantNode.test.js,74]     assert.deepEqual(expr.eval(), new bigmath.type.BigNumber(2.3));
[C,test/expression/node/ConstantNode.test.js,75]   });
[C,test/expression/node/ConstantNode.test.js,76] 
[C,test/expression/node/ConstantNode.test.js,77]   it ('should find a ConstantNode', function () {
[C,test/expression/node/ConstantNode.test.js,78]     var a = new ConstantNode('2', 'number');
[C,test/expression/node/ConstantNode.test.js,79]     assert.deepEqual(a.filter(function (node) {return node instanceof ConstantNode}),  [a]);
[C,test/expression/node/ConstantNode.test.js,80]     assert.deepEqual(a.filter(function (node) {return node instanceof SymbolNode}), []);
[C,test/expression/node/ConstantNode.test.js,81]   });
[C,test/expression/node/ConstantNode.test.js,82] 
[A,test/expression/node/ConstantNode.test.js,83]   it ('should throw an error when compiling an invalid value', function () {
[A,test/expression/node/ConstantNode.test.js,84]     var clone = math.create();
[A,test/expression/node/ConstantNode.test.js,85]     clone.config({number: 'number'});
[A,test/expression/node/ConstantNode.test.js,86]     assert.throws(function () { new ConstantNode('console.log("foo")', 'number').compile() }, /Invalid numeric value/)
[A,test/expression/node/ConstantNode.test.js,87]     clone.config({number: 'BigNumber'});
[A,test/expression/node/ConstantNode.test.js,88]     assert.throws(function () { new ConstantNode('console.log("foo")', 'number').compile() }, /Invalid numeric value/)
[A,test/expression/node/ConstantNode.test.js,89]     clone.config({number: 'Fraction'});
[A,test/expression/node/ConstantNode.test.js,90]     assert.throws(function () { new ConstantNode('console.log("foo")', 'number').compile() }, /Invalid numeric value/)
[A,test/expression/node/ConstantNode.test.js,91]   });
[A,test/expression/node/ConstantNode.test.js,92] 
[A,test/expression/node/ConstantNode.test.js,93]   it ('should escape quotes in strings', function () {
[A,test/expression/node/ConstantNode.test.js,94]     assert.strictEqual( new ConstantNode('"+foo+"', 'string').compile().eval.toString(),
[A,test/expression/node/ConstantNode.test.js,95]         'function (scope) {    if (scope) _validateScope(scope);    scope = scope || {};    return "\\"+foo+\\"";  }')
[A,test/expression/node/ConstantNode.test.js,96] 
[A,test/expression/node/ConstantNode.test.js,97]     assert.strictEqual( new ConstantNode('\\"escaped\\"', 'string').compile().eval.toString(),
[A,test/expression/node/ConstantNode.test.js,98]         'function (scope) {    if (scope) _validateScope(scope);    scope = scope || {};    return "\\"escaped\\"";  }')
[A,test/expression/node/ConstantNode.test.js,99]   });
[A,test/expression/node/ConstantNode.test.js,100] 
[A,test/expression/node/ConstantNode.test.js,101]   it ('should find a ConstantNode', function () {
[A,test/expression/node/ConstantNode.test.js,102]     var a = new ConstantNode('2', 'number');
[A,test/expression/node/ConstantNode.test.js,103]     assert.deepEqual(a.filter(function (node) {return node instanceof ConstantNode}),  [a]);
[A,test/expression/node/ConstantNode.test.js,104]     assert.deepEqual(a.filter(function (node) {return node instanceof SymbolNode}), []);
[A,test/expression/node/ConstantNode.test.js,105]   });
[A,test/expression/node/ConstantNode.test.js,106] 
[C,test/expression/node/ConstantNode.test.js,107]   it ('should run forEach on a ConstantNode', function () {
[C,test/expression/node/ConstantNode.test.js,108]     var a = new ConstantNode(2);
[C,test/expression/node/ConstantNode.test.js,109]     a.forEach(function () {
[C,test/expression/node/ConstantNode.test.js,110]       assert.ok(false, 'should not execute, constant has no childs')
[C,test/expression/node/ConstantNode.test.js,111]     });
[C,test/expression/node/ConstantNode.test.js,112]   });
[C,test/expression/node/ConstantNode.test.js,113] 
[C,test/expression/node/ConstantNode.test.js,114]   it ('should map a ConstantNode', function () {
[C,test/expression/node/ConstantNode.test.js,115]     var a = new ConstantNode(2);
[C,test/expression/node/ConstantNode.test.js,116]     var b = a.map(function () {
[C,test/expression/node/ConstantNode.test.js,117]       assert.ok(false, 'should not execute, constant has no childs')
[C,test/expression/node/ConstantNode.test.js,118]     });
[C,test/expression/node/ConstantNode.test.js,119] 
[C,test/expression/node/ConstantNode.test.js,120]     assert.notStrictEqual(b, a);
[C,test/expression/node/ConstantNode.test.js,121]     assert.deepEqual(b, a);
@@ -185,61 +185,61 @@ describe('security', function () {
[C,test/expression/security.test.js,185]     }, /Error: No access to property "value/);
[C,test/expression/security.test.js,186]   })
[C,test/expression/security.test.js,187] 
[C,test/expression/security.test.js,188]   it ('should not allow inserting fake nodes with bad code via node.map or node.transform', function () {
[C,test/expression/security.test.js,189]     assert.throws(function () {
[C,test/expression/security.test.js,190]       math.eval("badValue = {\"isNode\": true, \"_compile\": eval(\"f(a, b) = \\\"eval\\\"\")}; x = eval(\"f(child, path, parent) = path ==\\\"value\\\" ? newChild : child\", {\"newChild\": badValue}); parse(\"x = 1\").map(x).compile().eval()(\"console.log(\'hacked\')\")")
[C,test/expression/security.test.js,191]     }, /Error: Cannot compile node: unknown type "undefined"/);
[C,test/expression/security.test.js,192] 
[C,test/expression/security.test.js,193]     assert.throws(function () {
[C,test/expression/security.test.js,194]       math.eval("badValue = {\"isNode\": true, \"type\": \"ConstantNode\", \"valueType\": \"string\", \"_compile\": eval(\"f(a, b) = \\\"eval\\\"\")}; x = eval(\"f(child, path, parent) = path ==\\\"value\\\" ? newChild : child\", {\"newChild\": badValue}); parse(\"x = 1\").map(x).compile().eval()(\"console.log(\'hacked\')\")")
[C,test/expression/security.test.js,195]     }, /TypeError: fn is not a function/); // The error message is vague but well...
[C,test/expression/security.test.js,196]   })
[C,test/expression/security.test.js,197] 
[C,test/expression/security.test.js,198]   it.skip ('should not allow injecting code via an adjusted ConstantNode value', function () {
[C,test/expression/security.test.js,199]     assert.throws(function () {
[D,test/expression/security.test.js,200]         // TODO
[A,test/expression/security.test.js,200]       // TODO: write unit test
[C,test/expression/security.test.js,201]       }, /Error: No access to property "constructor/);
[C,test/expression/security.test.js,202]   })
[C,test/expression/security.test.js,203] 
[C,test/expression/security.test.js,204]   it.skip ('should not allow injecting code via an adjusted SymbolNode name', function () {
[C,test/expression/security.test.js,205]     assert.throws(function () {
[D,test/expression/security.test.js,206]         // TODO
[A,test/expression/security.test.js,206]       // TODO: write unit test
[C,test/expression/security.test.js,207]       }, /Error: No access to property "constructor/);
[C,test/expression/security.test.js,208]   })
[C,test/expression/security.test.js,209] 
[C,test/expression/security.test.js,210]   it.skip ('should not allow injecting code via an adjusted FunctionNode name', function () {
[C,test/expression/security.test.js,211]     assert.throws(function () {
[D,test/expression/security.test.js,212]         // TODO
[A,test/expression/security.test.js,212]       // TODO: write unit test
[C,test/expression/security.test.js,213]       }, /Error: No access to property "constructor/);
[C,test/expression/security.test.js,214]   })
[C,test/expression/security.test.js,215] 
[C,test/expression/security.test.js,216]   it.skip ('should not allow injecting code via an adjusted function name', function () {
[C,test/expression/security.test.js,217]     assert.throws(function () {
[D,test/expression/security.test.js,218]         // TODO
[A,test/expression/security.test.js,218]         // TODO: write unit test
[C,test/expression/security.test.js,219]       }, /Error: No access to property "constructor/);
[C,test/expression/security.test.js,220]   })
[C,test/expression/security.test.js,221] 
[C,test/expression/security.test.js,222]   it.skip ('should not allow injecting code via a property name', function () {
[C,test/expression/security.test.js,223]     assert.throws(function () {
[C,test/expression/security.test.js,224]         // TODO
[C,test/expression/security.test.js,225]       }, /Error: No access to property "constructor/);
[C,test/expression/security.test.js,226]   })
[C,test/expression/security.test.js,227] 
[C,test/expression/security.test.js,228]   it.skip ('should not allow inserting a fake ObjectNode', function () {
[C,test/expression/security.test.js,229]     assert.throws(function () {
[D,test/expression/security.test.js,230]         // TODO
[A,test/expression/security.test.js,230]       // TODO: write unit test
[C,test/expression/security.test.js,231]       }, /Error: No access to property "constructor/);
[C,test/expression/security.test.js,232]   })
[C,test/expression/security.test.js,233] 
[C,test/expression/security.test.js,234]   it ('should allow calling functions on math', function () {
[C,test/expression/security.test.js,235]     assert.equal(math.eval('sqrt(4)'), 2);
[C,test/expression/security.test.js,236]   })
[C,test/expression/security.test.js,237] 
[C,test/expression/security.test.js,238]   it ('should allow accessing properties on an object', function () {
[C,test/expression/security.test.js,239]     assert.deepEqual(math.eval('obj.a', {obj: {a:42}}), 42);
[C,test/expression/security.test.js,240]   })
[C,test/expression/security.test.js,241] 
[C,test/expression/security.test.js,242]   it ('should not allow accessing inherited properties on an object', function () {
[C,test/expression/security.test.js,243]     assert.throws(function () {
[C,test/expression/security.test.js,244]       math.eval('obj.constructor', {obj: {a:42}});
[C,test/expression/security.test.js,245]     }, /Error: No access to property "constructor"/)